### PR TITLE
[feat] Clean up bin/dev, move pipfile, and add devbox versioning

### PR DESCRIPTION
**Pull request by EpicPi**
## Changes made
- The `bin/dev` script used to find the running container and then run commands against it. This works but is apparently not the standard way of doing this. Instead, the script now takes advantage of docker-compose run to accomplish all the tasks.
   - This gives us the benefit of not having to have the entire app running if you just want to run `bin/dev check` for example.
- Move pipfile to the top level. This makes things cleaner.
- Move dev.env file to top level. Same reason
- Add versioning to devbox
  - Makes it easier during development of new versions
  - Allows you to work on old branches without having to rebase 

(Sorry, this probably should've been 4 prs.)

## Testing
- [ ] End-to-End

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -84,3 +84,12 @@ repos:
         args:
         -   "--req client/package.json"
         -   "--lock client/package-lock.json"
+-   repo: https://github.com/EpicPi/ReqLockHook
```
Comment (EpicPi) [inline .pre-commit-config.yaml:87]: Ensures that you update the version if you update the Dockerfile.

File: .dockerignore [hunk 2]
```diff
@@ -6,5 +6,5 @@ client
 .github
 .git
 bin
-server/.venv
+**/.venv
```
Comment (EpicPi) [inline .dockerignore:9]: Since we moved the pipfile, we moved the .venv location.

File: Dockerfile [hunk 3]
```diff
@@ -29,18 +29,19 @@ ENV MYSQL_USER user
 ENV MYSQL_PASSWORD password
 
 WORKDIR /home/worker/app/
-ENTRYPOINT bash
+CMD cd server && pipenv run flask run --host 0.0.0.0
```
Comment (EpicPi) [inline Dockerfile:32]: CMD allows you to override it when doing `docker run` ENTRYPOINT does not.

File: README.md [hunk 4]
```diff
@@ -18,18 +18,15 @@
 ```shell
 bin/dev setup
 ```
-5. Start python and mongo containers.
```
Comment (EpicPi) [inline README.md:21]: `docker-compose run` alleviates the need to do this. Instead you can just launch the ad hoc container for the task you want to run.

File: bin/dev [hunk 5]
```diff
@@ -1,63 +1,46 @@
...
+DEVBOX_VERSION=$(cat DEVBOX_VERSION)
 
-# Run commands in exiting dev container as correct user
-# args: commands to run in container
-runDockerExec(){
-    findBackendContainerId container_id
-    cmd="docker exec -it -u $(id -u):$(id -g) -w /home/worker/app/server $container_id "$@
-    bin/util logAndRun $cmd
-}
+DOCKER_COMPOSE="export DEVBOX_VERSION=$DEVBOX_VERSION && docker-compose"
```
Comment (EpicPi) [inline bin/dev:5]: Can't perform arbitrary shell commands in yaml, but can read the env variables, so explicitly set it before running `docker-compose`.

File: DEVBOX_VERSION [hunk 6]
```diff
@@ -0,0 +1 @@
+1
```
Comment (SammyBaek) [inline DEVBOX_VERSION:1]: Is this integer increment? Can you do semantic versioning?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (SammyBaek) [reply .pre-commit-config.yaml:87]: So `DEVBOX_VERSION` is akin to `package.json`'s `package-lock.json` for us? Does it get auto-generated or would we have to manually increase the version anytime we change the dockerfile

File: DEVBOX_VERSION (hunk 6 above)
Comment (EpicPi) [reply DEVBOX_VERSION:1]: It's just any string. We can do semver.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (EpicPi) [reply .pre-commit-config.yaml:87]: Manual increase. I just reused the gh action here because it did almost the exact thing we needed.

Comment (SammyBaek) [reply .pre-commit-config.yaml:87]: sgtm

File: DEVBOX_VERSION (hunk 6 above)
Comment (SammyBaek) [reply DEVBOX_VERSION:1]: ooo could we do that?

File: README.md [hunk 7]
```diff
@@ -57,3 +55,18 @@ bin/dev jstest [files]
...
+
+1. Occasionally, newer versions of the devobx container will be made. Your invocation of `bin/dev backend` will likely pull it, but you can manually do so by running.
+```shell
+bin/dev pull
+```
+
+## Fix Python Location for vscode
+
+```shell
+bin/dev run bash
```
Comment (EpicPi) [inline README.md:68]: I gave up trying to make this work in bash script. You'll just have to copy paste couple lines if you want it.

**Review by SammyBaek — APPROVED** [review]
lgtm!

**Pull request merged**